Fix compilation of serialisation code on 32 bit platforms

SquidDev opened the pull request:
The `Hashable Node` implementation uses a 64-bit hash. However, this does not work on 32 bit platforms, such as WASM, as the 64-bit constants do not fit into a 32 bit word.

Ideally we'd be able to keep the current algorithm and use `Word64`. However, that's awkward to do on older versions of GHC (as we'd have to depend on `ghc-prim`), so instead we just use different constants.

SquidDev committed:
Fix compilation of serialisation code on 32 bit platforms

The "Hashable Node" implementation uses a 64-bit hash. However, this
does not work on 32 bit platforms (as the 64-bit constants do not fit
into a 32 bit word).

This commit uses alternative constants when on 32-bit platforms. The
hash factor is the 32-bit fibonacci constant, and the salt is
arbitrarily chosen.

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Base.hs:
```diff
@@ -67,13 +69,23 @@ instance Hashable Node where
+    factor = 11400714819323198549
+#else
+    factor = 2654435741
+#endif
```
Source for the magic numbers: https://stackoverflow.com/questions/4113278/64-bit-multiplicative-hashing

SquidDev committed:
Explain the source of some of the constants

The salt really does just seem to be "I made it up". It's copied and
pasted several places, but just goes back to Andras!

nad commented:
Does this mean that one cannot take a `.agdai` file that is produced on a 64-bit system and use on a 32-bit system, or vice versa? I think the `.agdai` files should be platform-independent.

SquidDev commented:
No, that still works! The hash function is only used for deduplicating values during serialisation, it's never used as part of the file contents.
